pkg/pool-stable-phantom/test/StablePhantomPool.test.ts

```diff
@@ -641,17 +640,160 @@ describe('StablePhantomPool', () => {
       });
     });
 
-    describe('exit', () => {
-      sharedBeforeEach('deploy pool', async () => {
-        await deployPool();
-        await pool.init({ recipient, initialBalances });
+    describe('onExitPool', () => {
+      let previousBptBalance: BigNumber;
+
+      sharedBeforeEach('deploy and initialize pool', async () => {
+        await deployPool({ admin });
+        await pool.init({ initialBalances, recipient: lp });
+        previousBptBalance = await pool.balanceOf(lp);
       });
 
-      context('when the sender is not the vault', () => {
-        it('reverts', async () => {
-          const tx = pool.instance.onExitPool(pool.poolId, ZERO_ADDRESS, ZERO_ADDRESS, [0], 0, 0, '0x');
-          await expect(tx).to.be.revertedWith('CALLER_NOT_VAULT');
+      sharedBeforeEach('allow vault', async () => {
+        await tokens.mint({ to: lp, amount: fp(100) });
+        await tokens.approve({ from: lp, to: pool.vault });
+      });
+
+      it('fails if caller is not the vault', async () => {
+        await expect(
+          pool.instance.connect(lp).onExitPool(pool.poolId, recipient.address, other.address, [0], 0, 0, '0x')
+        ).to.be.revertedWith('CALLER_NOT_VAULT');
+      });
+
+      it('fails if no user data', async () => {
+        await expect(pool.exit({ data: '0x' })).to.be.revertedWith('Transaction reverted without a reason');
+      });
+
+      it('fails if wrong user data', async () => {
+        const wrongUserData = ethers.utils.defaultAbiCoder.encode(['address'], [lp.address]);
+
+        await expect(pool.exit({ data: wrongUserData })).to.be.revertedWith('Transaction reverted without a reason');
+      });
+
+      describe('exit exact BPT in for all tokens out', () => {
+        context('not in recovery mode', () => {
+          itExitsExactBptInForAllTokensOutProperly();
+        });
+
+        context('in recovery mode', () => {
+          sharedBeforeEach('enable recovery mode', async () => {
+            await pool.enableRecoveryMode(admin);
+          });
+
+          itExitsExactBptInForAllTokensOutProperly();
         });
+
+        function itExitsExactBptInForAllTokensOutProperly() {
+          it('grants all tokens for exact bpt', async () => {
+            // Exit with half of the BPT balance
+            const bptIn = previousBptBalance.div(2);
+            const expectedAmountsOut = initialBalances.map((balance) => bn(balance).div(2));
+
+            const result = await pool.multiExitGivenIn({ from: lp, bptIn });
+
+            // Balances are reduced by half because we are returning half of the BPT supply
+            expect(result.amountsOut).to.be.equalWithError(expectedAmountsOut, 0.001);
+
+            // Current BPT balance should have been reduced by half
+            expect(await pool.balanceOf(lp)).to.be.equalWithError(bptIn, 0.001);
```

Comment:
```suggestion
            // Exit with a third of the BPT balance
            const bptIn = previousBptBalance.div(3);
            // The pool should return about a third of its holdings
            const expectedAmountsOut = initialBalances.map((balance) => bn(balance).div(3));

            const result = await pool.multiExitGivenIn({ from: lp, bptIn });

            expect(result.amountsOut).to.be.equalWithError(expectedAmountsOut, 0.001);

            // Current BPT balance should have been reduced to two thirds
            expect(await pool.balanceOf(lp)).to.equa(previousBptBalance.sub(bptIn));
```

Avoid using 50% amounts - also changed a test to check exact amounts since there should be no error in that one.